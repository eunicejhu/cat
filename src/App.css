@import "./styles/validate.css";

html {
  height: 100%;
}
body {
  margin: 0;
  color: #e3e4eb;
  background: linear-gradient(
    45deg,
    rgba(66, 183, 245, 0.8) 0%,
    rgba(66, 245, 189, 0.4) 100%
  );
  font: 600 16px/18px "Open Sans", sans-serif;
}
*,
:after,
:before {
  box-sizing: border-box;
}
.clearfix:after,
.clearfix:before {
  content: "";
  display: table;
}
.clearfix:after {
  clear: both;
  display: block;
}
a {
  color: inherit;
  text-decoration: none;
}
