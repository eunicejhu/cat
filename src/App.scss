@import "./styles/validate.css";

html {
  height: 100%;
}
body {
  margin: 0;
  color: #4eade4;
  background: #e3e4eb;
  font: 600 16px/18px "Open Sans", sans-serif;
  padding: 0 8em;
}
*,
:after,
:before {
  box-sizing: border-box;
}
.clearfix:after,
.clearfix:before {
  content: "";
  display: table;
}
.clearfix:after {
  clear: both;
  display: block;
}
a {
  color: inherit;
  text-decoration: none;
}

#root {
  height: 100vh;
}

nav {
  display: flex;
  height: 4em;
  line-height: 4em;

  li {
    list-style: none;
    margin: 0 10px;

    &:hover {
      cursor: pointer;
    }
  }
}
main {
  position: relative;
  min-height: 70%;
}

footer {
  margin: 3em 0;
  line-height: 3em;
}

.add-post-form {
  width: 100%;
  div {
    width: 100%;
  }
}

form {
  display: flex;
  flex-direction: column;
  align-items: center;
}
div.field {
  display: flex;
  padding: 10px 0;
  justify-content: center;
  width: 100%;
  label {
    width: 6em;
    display: inline-block;
  }
  input,
  textarea,
  select {
    width: 80%;
    border: none;
    border-radius: 4px;
  }
  input,
  select {
    height: 2em;
    outline: none;
  }
  textarea {
    height: 8em;
    outline: none;
  }
  button,
  input[type="button"] {
    padding: 0.6em 1.2em;
    border-radius: 4px;
    width: 9em;
    &:hover {
      cursor: pointer;
      font-weight: bold;
    }
  }
}

section {
  display: flex;
  flex-direction: column;
  align-items: center;

  h2 {
    margin: 2em;
  }

  article {
    width: 100%;
    margin: 20px;
    border: 1px solid;
    padding: 20px;
    border-radius: 4px;
    box-shadow: 2px 2px;
  }
  .reactions-container {
    button {
      outline: none;
      margin-right: 1em;
      border: 0px;
      border-radius: 26px;
      padding: 4px 16px;
      &:hover {
        cursor: pointer;
        background-color: #4eade4;
      }
    }
  }
}
