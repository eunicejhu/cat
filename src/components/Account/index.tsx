import React from "react";
import PasswordChangeForm from "../PasswordChange";
import { PasswordForgetForm } from "../PasswordForget";
import { StyledPage } from "../Page";
import { useAuthorization } from "../Session";
const Account = () => {
    const authUser = useAuthorization();
    return authUser ? (
        <StyledPage>
            <h2>Reset By Email</h2>
            <PasswordForgetForm />
            <PasswordChangeForm />
        </StyledPage>
    ) : null;
};

export default Account;
