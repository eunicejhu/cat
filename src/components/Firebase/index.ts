import app from "firebase/app";
import "firebase/auth";
import FirebaseContext from "./context";
import useFirebase from "./useFirebase";

//TODO: .env.production, .env.development
const config = {
    apiKey: process.env.REACT_APP_API_KEY,
    authDomain: process.env.REACT_APP_AUTH_DOMAIN,
    databaseURL: process.env.REACT_APP_DATABASE_URL,
    projectId: process.env.REACT_APP_PROJECT_ID,
    storageBucket: process.env.REACT_APP_STORAGE_BUCKET,
    messagingSenderId: process.env.REACT_APP_MESSAGING_SENDER_ID,
    appId: process.env.REACT_APP_APP_ID,
    measurementId: process.env.REACT_APP_MEASUREMENT_ID,
};

class Firebase {
    auth: app.auth.Auth;
    constructor() {
        app.initializeApp(config);
        this.auth = app.auth();
        this.auth.useEmulator("http://localhost:9099/");
    }

    /**
     * Auth API
     */
    doCreateUserWithEmailAndPassword = (email: string, password: string) =>
        this.auth.createUserWithEmailAndPassword(email, password);
    doSignInWithEmailAndPassword = (email: string, password: string) =>
        this.auth.signInWithEmailAndPassword(email, password);
    doSignOut = () => this.auth.signOut();
    doPasswordReset = (email: string) =>
        this.auth.sendPasswordResetEmail(email);
    doPasswordUpdate = (newPassword: string) =>
        this.auth.currentUser?.updatePassword(newPassword);
}

export default Firebase;
export { FirebaseContext, useFirebase };
