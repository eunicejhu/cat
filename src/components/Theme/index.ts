import ThemeContext, { themes } from "./Context";
import useTheme from "./useTheme";

export type { Mode, Themes, SetMode } from "./Context";
export { ThemeContext, themes, useTheme };
