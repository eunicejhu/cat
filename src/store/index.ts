import { configureStore } from "@reduxjs/toolkit";
import postsReducer, { Post } from "../features/posts/postsSlice";
import usersReducer from "../features/users/usersSlice";
import notificationsReducer, {
  Notification,
} from "../features/notifications/notificationsSlice";
import { useDispatch } from "react-redux";

const store = configureStore({
  reducer: {
    posts: postsReducer,
    users: usersReducer,
    notifications: notificationsReducer,
  },
});

export type AppDispatch = typeof store.dispatch;
export const useAppDispatch = () => useDispatch<AppDispatch>();

export type State = ReturnType<typeof store.getState>;
// selectors
export const selectAllPosts = (state: State): Post[] => state.posts.data;
export const selectPostById = (
  state: State,
  postId: string
): Post | undefined => state.posts.data.find((post) => post.id === postId);
export const selectAllNotifications = (state: State): Notification[] =>
  state.notifications;

export default store;
