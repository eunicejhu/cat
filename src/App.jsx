import React, { useState } from "react";
import { BrowserRouter as Router } from "react-router-dom";
import { CookiesProvider } from "react-cookie";
import history from "./helpers/history";
import PortfolioPage from "./pages/PortfolioPage";
import ThemeContext, { themes } from "./context/ThemeContext";

import "./styles/App.css";

const App = () => {
  const [theme, setTheme] = useState(themes.dark);
  return (
    <ThemeContext.Provider value={{ theme, setTheme }}>
      <CookiesProvider>
        <Router history={history}>
          <PortfolioPage />
        </Router>
      </CookiesProvider>
    </ThemeContext.Provider>
  );
};
export default App;
