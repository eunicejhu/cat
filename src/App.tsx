import React, { useEffect, useState } from "react";
import NavBar from "./NavBar";
import Routes from "./Routes";
import styled from "styled-components";
import { Mode, Themes, SetMode } from "./components/Theme";
import Select from "./components/select/Select";
import app from "firebase/app";

import { useDispatch } from "react-redux";
import { fetchUsers } from "./features/users/usersSlice";
import Firebase, { useFirebase } from "./components/Firebase";

export interface AppProps {
    themes: Themes;
    mode: Mode;
    setMode: SetMode;
}

const StyledApp = styled.div<AppProps>`
    ${({ themes, mode }) => {
        return `
            color: ${themes[mode].text};
            background-color: ${themes[mode].background};
      
        & input, & textarea {
            border: 1px solid  ${themes[mode].btnBorder};
            padding:  ${themes[mode].spacing};
            box-sizing: border-box;
        }

        & header {
            display: flex;
            align-items: center;
            justify-content: flex-end;

            nav {
                display: flex;
                align-items: center;
                li {
                    margin: 0 ${themes[mode].spacing};
                    list-style: none;
                }
            }

            .selectWrapper {
                width: 4em;
                margin: 0 ${themes[mode].spacing};
            }
        }
        `;
    }}
    : var(--color-text);
`;

const modeData = (data: string[]) =>
    data.map((mode: string) => ({ id: mode, name: mode }));

export const useAuthUser = () => {
    const [authUser, setAuthUser] = useState<app.User | null>(null);
    const firebase = useFirebase() as Firebase;
    useEffect(() => {
        // unsubscribe is reported to be "not a function" weird!
        firebase.auth.onAuthStateChanged((authUser) => setAuthUser(authUser));
    }, [firebase]);

    return authUser;
};

const App: React.FC<AppProps> = (props) => {
    const dispatch = useDispatch();
    const authUser = useAuthUser();

    const { themes, mode, setMode } = props;
    useEffect(() => {
        dispatch(fetchUsers());
    }, [dispatch]);

    return (
        <StyledApp {...props} themes={themes} mode={mode}>
            <header>
                <NavBar authUser={authUser} />
                <div className="selectWrapper">
                    <Select
                        themes={themes}
                        mode={mode}
                        value={mode}
                        name="theme"
                        data={modeData(["dark", "pink", "light"])}
                        placeholder="theme"
                        onChange={setMode}
                    ></Select>
                </div>
            </header>
            <main>
                <Routes />
            </main>
            <footer>copyright@2020 author ZUOQIN HU</footer>
        </StyledApp>
    );
};

export default App;
