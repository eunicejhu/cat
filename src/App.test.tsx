import React from "react";
import { render } from "@testing-library/react";
import App, { useAuthUser } from "./App";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import store from "./store/index";
import Firebase, { FirebaseContext } from "../src/components/Firebase";

import { themes } from "./components/Theme";
import { renderHook } from "@testing-library/react-hooks";
let mockOnAuthStateChanged = jest.fn().mockReturnValue(jest.fn());
jest.mock("../src/components/Firebase", () => ({
    __esModule: true,
    ...(jest.requireActual("../src/components/Firebase") as object),
    default: function () {
        return {
            auth: {
                onAuthStateChanged: mockOnAuthStateChanged,
            },
        };
    },
}));
describe("App", () => {
    it("useAuthUser", () => {
        const firebase = new Firebase();
        const wrapper = ({ children }: { children: React.ReactNode }) => (
            <FirebaseContext.Provider value={firebase}>
                {children}
            </FirebaseContext.Provider>
        );
        renderHook(() => useAuthUser(), { wrapper });
        expect(mockOnAuthStateChanged).toHaveBeenCalledTimes(1);
    });

    it("render correctly", () => {
        const ui = (
            <FirebaseContext.Provider value={new Firebase()}>
                <Provider store={store}>
                    <BrowserRouter>
                        <App
                            themes={themes}
                            mode={"pink"}
                            setMode={jest.fn()}
                        />
                    </BrowserRouter>
                </Provider>
            </FirebaseContext.Provider>
        );
        const { asFragment } = render(ui);
        expect(asFragment()).toMatchSnapshot();
    });
});
