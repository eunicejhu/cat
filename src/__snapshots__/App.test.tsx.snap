// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-fubCfw Wzshu"
    mode="pink"
  >
    <header>
      <nav>
        <li>
          <a
            class="sc-gsTCUz flwUht"
            href="/"
            mode="pink"
            themes="[object Object]"
          >
            Landing
          </a>
        </li>
        <li>
          <a
            class="sc-gsTCUz flwUht"
            href="/signin"
            mode="pink"
            themes="[object Object]"
          >
            Sign In
          </a>
        </li>
      </nav>
      <div
        class="selectWrapper"
      >
        <div
          class="sc-hKgILt javMHI"
          mode="pink"
          name="theme"
        >
          <input
            class="selectInput"
            data-testid="selectInput"
            placeholder="theme"
            type="text"
            value="pink"
          />
          <ul
            class="close"
          >
            <li
              class="selectInput-li"
              data-id="dark"
              data-testid="dark"
            >
              dark
            </li>
            <li
              class="selectInput-li"
              data-id="pink"
              data-testid="pink"
            >
              pink
            </li>
            <li
              class="selectInput-li"
              data-id="light"
              data-testid="light"
            >
              light
            </li>
          </ul>
        </div>
      </div>
    </header>
    <main>
      <section
        class="add-post-form"
      >
        <form
          class="sc-eCssSg eTymvH"
        >
          <div
            class="field"
          >
            <label
              for="title"
            >
              Title
            </label>
            <input
              data-testid="title"
              id="title"
              name="title"
              value=""
            />
          </div>
          <div
            class="field"
          >
            <label
              for="content"
            >
              Content
            </label>
            <textarea
              data-testid="content"
              id="content"
              name="content"
            />
          </div>
          <div
            class="field"
          >
            <label
              for="users"
            >
              Author
            </label>
            <div>
              <div
                class="sc-hKgILt javMHI"
                data-testid="users"
                id="users"
                mode="pink"
              >
                <input
                  class="selectInput"
                  data-testid="selectInput"
                  placeholder="choose the user"
                  type="text"
                  value=""
                />
                <ul
                  class="close"
                />
              </div>
            </div>
          </div>
          <div
            class="field"
          >
            <button
              class="sc-dlfnbm eTNEQD"
              disabled=""
              mode="pink"
              type="button"
            >
              Add Post
            </button>
            <span />
          </div>
        </form>
      </section>
      <section
        class="posts-list"
      >
        <h2>
          Posts
        </h2>
        <div>
          loading...
        </div>
      </section>
    </main>
    <footer>
      copyright@2020 author ZUOQIN HU
    </footer>
  </div>
</DocumentFragment>
`;
